### PR TITLE
test-network: fixlets and performance improvements

### LINKED ISSUE #23964: `test_dhcp_client_with_ipv4ll` from the systemd-networkd test suite is unstable
### systemd version the issue has been seen with

latest main

### Used distribution

Arch Linux

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

_No response_

### Component

tests

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

Looks like the test needs bit of tweaking after the recent changes:

```
======================================================================
FAIL: test_dhcp_client_with_ipv4ll (__main__.NetworkdDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/./test/test-network/systemd-networkd-tests.py", line 4940, in test_dhcp_client_with_ipv4ll
    self.wait_address_dropped('veth99', r'inet 192\.168\.5\.\d+/24 metric 1024 brd 192\.168\.5\.255 scope global dynamic', ipv='-4', timeout_sec=120)
  File "/build/./test/test-network/systemd-networkd-tests.py", line 726, in wait_address_dropped
    self.assertNotRegex(output, address_regex)
AssertionError: Regex matched: 'inet 192.168.5.181/24 metric 1024 brd 192.168.5.255 scope global dynamic' matches 'inet 192\\.168\\.5\\.\\d+/24 metric 1024 brd 192\\.168\\.5\\.255 scope global dynamic' in '111: veth99@veth-peer: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000\n    inet 192.168.5.181/24 metric 1024 brd 192.168.5.255 scope global dynamic veth99\n       valid_lft 0sec preferred_lft 0sec'

----------------------------------------------------------------------
Ran 143 tests in 1778.904s

FAILED (failures=1, skipped=1, expected failures=3)
```

https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/12784/artifact//systemd-centos-ci/artifacts_5zth3lke/vagrant-logs.NXL/vagrant-arch-testsuite.Igu/systemd-networkd-tests.py_FAIL.log

```
======================================================================
FAIL: test_dhcp_client_with_ipv4ll (__main__.NetworkdDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/./test/test-network/systemd-networkd-tests.py", line 4940, in test_dhcp_client_with_ipv4ll
    self.wait_address_dropped('veth99', r'inet 192\.168\.5\.\d+/24 metric 1024 brd 192\.168\.5\.255 scope global dynamic', ipv='-4', timeout_sec=120)
  File "/build/./test/test-network/systemd-networkd-tests.py", line 726, in wait_address_dropped
    self.assertNotRegex(output, address_regex)
AssertionError: Regex matched: 'inet 192.168.5.181/24 metric 1024 brd 192.168.5.255 scope global dynamic' matches 'inet 192\\.168\\.5\\.\\d+/24 metric 1024 brd 192\\.168\\.5\\.255 scope global dynamic' in '111: veth99@veth-peer: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000\n    inet 192.168.5.181/24 metric 1024 brd 192.168.5.255 scope global dynamic veth99\n       valid_lft 0sec preferred_lft 0sec'

----------------------------------------------------------------------
Ran 143 tests in 1777.767s

FAILED (failures=1, skipped=1, expected failures=3)
```

https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-vagrant-archlinux/12786/artifact//systemd-centos-ci/artifacts_76pna3xz/vagrant-logs.Ilh/vagrant-arch-testsuite.6ra/systemd-networkd-tests.py_FAIL.log

/cc @yuwata

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23964.

### CODE REVIEW

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -58,260 +53,546 @@
...
+def mkdir_p(path):
+    os.makedirs(path, exist_ok=True)
+
+def touch(path):
+    Path(path).touch()
+
+def check_output(*command, text=True, **kwargs):
+    # This checks the result and returns stdout on success.
+    command = command[0].split() + list(command[1:])
+    return subprocess.run(command, check=True, universal_newlines=text, capture_output=True, **kwargs).stdout.rstrip()
```
Comment (mrc0mmand) [inline test/test-network/systemd-networkd-tests.py]: Looks like `capture_output` was introduced in Python 3.7, and the C8S nodes in CentOS CI are on Python 3.6. Should I bump it to Python 3.8 or using `stdout=PIPE` and `stderr=PIPE` would be a sufficient alternative?

Comment (yuwata) [reply test/test-network/systemd-networkd-tests.py]: Thank you for catching that. That's not necessary, Now `std{out,err}=PIPE` are used.

**Comment by yuwata** [comment]
Yay! Much much faster now! PTAL.

| | issue #23723 | with PR #23950 | with PR #23959 | with this PR |
| --- | --- | --- | --- | --- |
| CentOS CI (CentOS Stream 8) | 41:49.07 | 37:03.35 (88.6%) | 33:04.21 (79.1%) | 23:57.51 (57.3%) |
| CentOS CI (Arch Linux) | 39:18.83 | 34:34.36 (87.9%) | 29:57.67 (76.2%) | 22:25.40 (57.0%) |
| CentOS CI (Arch Linux + sanitizers) | 42:25.91 (from #23949) | 34:15.61 (80.7%) | 32:19.12 (76.2%) | 23:54.21 (56.3%) |

**Comment by mrc0mmand** [comment]
> Yay! Much much faster now! PTAL.
> 	issue #23723 	with PR #23950 	with PR #23959 	with this PR
> CentOS CI (CentOS Stream 8) 	41:49.07 	37:03.35 (88.6%) 	33:04.21 (79.1%) 	23:57.51 (57.3%)
> CentOS CI (Arch Linux) 	39:18.83 	34:34.36 (87.9%) 	29:57.67 (76.2%) 	22:25.40 (57.0%)
> CentOS CI (Arch Linux + sanitizers) 	42:25.91 (from #23949) 	34:15.61 (80.7%) 	32:19.12 (76.2%) 	23:54.21 (56.3%)

Wow, that's impressive, thanks a lot for working on this!

**Review by mrc0mmand — APPROVED** [review]
LGTM, thanks!

**Pull request merged**